src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.Build.BackEnd.Logging;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.Build.BackEnd.Components.RequestBuilder
+{
+    internal class AssemblyLoadsTracker : MarshalByRefObject, IDisposable
+    {
+        private static readonly ConcurrentDictionary<AppDomain, AssemblyLoadsTracker> s_instances = new();
+        private readonly LoggingContext? _loggingContext;
+        private readonly LoggingService? _loggingService;
+        private readonly AssemblyLoadingContext _context;
+        private readonly AppDomain _appDomain;
+
+        private AssemblyLoadsTracker(
+            LoggingContext? loggingContext,
+            LoggingService? loggingService,
+            AssemblyLoadingContext context,
+            AppDomain appDomain)
+        {
+            _loggingContext = loggingContext;
+            _loggingService = loggingService;
+            _context = context;
+            _appDomain = appDomain;
+        }
+
+        public static IDisposable StartTracking(
+            LoggingContext loggingContext,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
+            return StartTracking(loggingContext, null, context, appDomain);
+        }
+
+        public static IDisposable StartTracking(
+            LoggingService loggingService,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
+            return StartTracking(null, loggingService, context, appDomain);
+        }
+
+        public static void StopTracking(AppDomain appDomain)
+        {
+            if (s_instances.TryRemove(appDomain, out AssemblyLoadsTracker? tracker))
+            {
+                tracker.StopTracking();
+            }
+        }
+
+        public void Dispose()
+        {
+            StopTracking();
+        }
+
+        private static IDisposable StartTracking(
+            LoggingContext? loggingContext,
+            LoggingService? loggingService,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
+            var tracker = new AssemblyLoadsTracker(loggingContext, loggingService, context, appDomain ?? AppDomain.CurrentDomain);
+            if (appDomain != null)
+            {
+                s_instances.AddOrUpdate(appDomain, tracker, (_, loadsTracker) => loadsTracker);
+            }
+            tracker.StartTracking();
```

Comment:
Note: consider the case where the same `AppDomain` is registered twice (the crux of my comment above). This code actually produces a subtle race condition. There is a period of time inside the call to `StartTracking` that assembly loads are not monitored and that means we could end up missing an entry in the log.

Reply:
This convo was marked as resolved but I can't see how. The race condition still exists.

Reply:
Thanks for reiterating this one!
Since the `StartTracking` instance method was private and was invoked only after construction - there was no chance for double subscription case - so the `unsubscribe-subscribe` was superfluous and I removed it. 
But it's a good point for future cases as well